config/typescript.js

```diff
@@ -7,10 +7,16 @@
 // Omit `.d.ts` because 1) TypeScript compilation already confirms that
 // types are resolved, and 2) it would mask an unresolved
 // `.ts`/`.tsx`/`.js`/`.jsx` implementation.
-const typeScriptExtensions = ['.ts', '.cts', '.mts', '.tsx'];
+const typeScriptKeys = ['ts', 'cts', 'mts', 'tsx'];
+const typeScriptExtensions = typeScriptKeys.map((key) => `.${key}`);
+const typeScriptFiles = typeScriptExtensions.map((ext) => `**/*${ext}`);
 
 const allExtensions = [...typeScriptExtensions, '.js', '.jsx'];
 
+const typeScriptRules = Object.fromEntries(
+  typeScriptKeys.map((key) => [key, 'never']),
```

Comment:
Doesn’t this depend on your resolution settings in tsconfig?

Reply:
In the most recent versions of TypeScript we can configure in the `tsconfig` if we want `ts` extensions to be allowed or not while importing (https://www.typescriptlang.org/tsconfig#allowImportingTsExtensions). 

That's why it's good to deactivate this eslint rule for `ts` extensions, so the tsconfig can raise an error or not without conflicting with it.

Reply:
then it shouldn't be setting it to "never", should it? We don't have an "ignore" setting, so i don't think adding this makes much sense.

Reply:
You're right, but `never` is the default and main use of typescript (compilation), the other use is type-checking-only. 

In my opinion a user setting the eslint-import-plugin typescript config wouldn't like to constantly have an eslint issue although he respects the typescript rules.

Reply:
I agree - which is why the override here should be _disabling_ the rule, not configuring it.

Reply:
however, what does the typescript-eslint plugin disable? is this rule not already on the list?

Reply:
I don't see it configured: https://github.com/typescript-eslint/typescript-eslint/tree/d2973cca391eeb763d72c5ef2f49cdafe4ffc216/packages/eslint-plugin/src/configs

Reply:
They just make a recommendation of use or not, indeed depending on `modeResolution` (https://github.com/typescript-eslint/typescript-eslint/blob/d2973cca391eeb763d72c5ef2f49cdafe4ffc216/docs/linting/troubleshooting/Performance.mdx#L94)

Reply:
ok - it sounds like unless we read the tsconfig here, we can't actually configure OR disable this rule, and safely cover all use cases?

Reply:
You're right.

Reply:
> I agree - which is why the override here should be _disabling_ the rule, not configuring it.

Disabled by the new commit

Reply:
right, but per the typescript-eslint recommendation, we can't just disable it either.

Reply:
Ok, so should depend on `modeResolution`

Reply:
@ljharb, and what about a new entry point `typescript-node16` importing current `typescript` config and adding `overrides` to it?

Reply:
a new entry point may be reasonable, but calling it "node16" implies you should be using it if you're on node 16+, which isn't the case.

instead could we use tsconfig-paths to read the tsconfig and dynamically set the right thing?

Reply:
I'm not familiar with but I can try. I need to have in mind that sometimes a tsconfig extends another one beforehand.

Reply:
True, but we don't support that since we're stuck on tsconfig-paths v3, and they didn't backport that feature from v4. It would be automatically handled if that changes, so you don't have to worry about it.